core/src/main/kotlin/model/Documentable.kt

```diff
@@ -5,231 +5,273 @@ import org.jetbrains.dokka.model.doc.DocumentationNode
+}
+
+interface withType {
+    val type: PlatformDependent<TypeWrapper>
+}
+
+interface withAbstraction {
+    val isAbstract: PlatformDependent<Boolean>
+    val isOpen: PlatformDependent<Boolean>
+    val isFinal: PlatformDependent<Boolean>
+}
+
+interface withCompanion {
+    val companion: List<Object>
+}
+
+interface withConstructors {
+    val constructors: List<Function>
+}
+
+interface withGenerics {
+    val generics: PlatformDependent<Classlike>
+}
+
+interface Callable : withVisibility, withType, withAbstraction, withExpectActual {
+    val receiver: PlatformDependent<Parameter>
+}
+
+interface CanBeSupertype
+
+interface Classlike : withScope, withVisibility, withExpectActual, CanBeSupertype {
+    val supertypes: PlatformDependent<CanBeSupertype>
+}
+
+class Module(
+    override val name: String,
+    override val packages: List<Package>,
+    override val documentation: PlatformDependent<DocumentationNode>,
+    override val original: PlatformDependent<Module>
+) : Documentable(), withPackages {
     override val dri: DRI = DRI.topLevel
-    override val children: List<Package> = packages
-    override val extra: MutableSet<Extra> = mutableSetOf()
+    override val children: List<Documentable>
+        get() = packages
 }
 
 class Package(
     override val dri: DRI,
     override val functions: List<Function>,
     override val properties: List<Property>,
     override val classlikes: List<Classlike>,
-    override val extra: MutableSet<Extra> = mutableSetOf()
-) : ScopeNode() {
+    override val packages: List<Package>,
+    override val documentation: PlatformDependent<DocumentationNode>,
+    override val original: PlatformDependent<Package>
+) : Documentable(), withScope, withPackages {
     override val name = dri.packageName.orEmpty()
+    override val children: List<Documentable>
+        get() = (functions + properties + classlikes + packages) as List<Documentable>
 }
 
 class Class(
     override val dri: DRI,
     override val name: String,
-    override val kind: ClassKind,
-    val constructors: List<Function>,
+    override val constructors: List<Function>,
     override val functions: List<Function>,
     override val properties: List<Property>,
     override val classlikes: List<Classlike>,
-    override val expected: ClassPlatformInfo?,
-    override val actual: List<ClassPlatformInfo>,
-    override val extra: MutableSet<Extra> = mutableSetOf(),
-    override val visibility: Map<PlatformData, Visibility>
-) : Classlike(
-    name = name,
-    dri = dri,
-    kind = kind,
-    functions = functions,
-    properties = properties,
-    classlikes = classlikes,
-    expected = expected,
-    actual = actual,
-    extra = extra
-), WithVisibility
+    override val expect: Class?,
+    override val actual: PlatformDependent<Class>,
+    override val visibility: PlatformDependent<Visibility>,
+    override val companion: List<Object>,
+    override val generics: PlatformDependent<Classlike>,
```

Comment:
I don't think `ClassLike` gives us information we need here. Maybe `TypeWrapper`?

Reply:
Changed to `TypeWrapper` for now and added a TODO to fix this, preferably with PR https://github.com/Kotlin/dokka/issues/659